### ISSUE TITLE
Major migration problem

### ISSUE THREAD

**Issue by Lucki2g** [DevelOpsITU/MiniTwit#109]
I woke up this morning I suddenly realized we had not migrated the correct data. What I mean, is that the data we migrated Tuesday 8th March was all salts generated the new way, since @Thomsen96 and I sat on the branch which contains the new salt creation.

I redid the entire migration testing using following method

## How I tested the migration
- switch branch to dev (or any place where the salt creation haven't changed)
- start application and generate a few users
- save the .db file in the ../../data/ directory from ./scripts
SQLiteBrowser image of user table:
![image](https://user-images.githubusercontent.com/11597762/157462650-4c0310f4-38df-4067-be8e-bfcd2aadc13f.png)
- switch branch back to #92 
- run `docker compose up -d` in root directory
- run `docker ps` check that postgres:13.5 is there and copy id
- run `docker exec -it <id> bash`
- run `createdb -U postgres minitwit`
- run `exit`
- run `docker compose up -d`
- run `docker ps` check that minitwit-Postgres has been created
- add a new postgres server to whatever database management tool you use with these properties:
![image](https://user-images.githubusercontent.com/11597762/157466938-2eb1f8ee-bfa3-47d7-bb4f-7adfcdcef647.png)
- switch directory to scripts
- run `docker compose -f .\pgloader.docker-compose.yaml up -d`
**If you now inspect the database it has not transferred the users!!!**
This is the warning that the pgloader docker container prints: (visible with command `docker logs pgloader`)
![image](https://user-images.githubusercontent.com/11597762/157479202-b7194e62-874f-4677-a695-7aa05602fe85.png)
But more importantly it is visible that an error occurred on transfer of user table
![image](https://user-images.githubusercontent.com/11597762/157479246-5cf360d5-26a4-41cc-9373-923e83749399.png)
The error is visible a little higher up
![image](https://user-images.githubusercontent.com/11597762/157473172-d3fa1bfa-ae4c-44a5-99f8-d9ce497b92d5.png)
This was the error I was afraid of in [#92.](https://github.com/DevelOpsITU/MiniTwit/issues/92#issuecomment-1060429133)

I tried doing the whole chabang again but changed the gorm specifications for the User table to: 
![image](https://user-images.githubusercontent.com/11597762/157473483-af6a0860-0568-46de-801a-46752e2d9971.png)
(forgot to delete the docker image, so actually never did, however did similar later)
Read here [gorm documentation](https://gorm.io/docs/models.html) that database specific types has to be full CAPS
The type of the struct has to match figured out here [postgres for orm documentation](https://pg.uptrace.dev/models/)
- this means the type has to be []byte which is not really easy to change since we use the model in many places
I did the whole chabang once again... this time I remembered to delete the image and rebuild the code now with the datatype changed to []byte and gorm type to bytea. This resulted in same  UTF-8 error, but this warning confirming it tried to transfer the sqlite string password to the bytea type (postgres BLOB).
![image](https://user-images.githubusercontent.com/11597762/157477714-dad96af1-37a8-45e9-815b-f194996bab99.png)

## TLDR:
The migration of the users will not work, because the invalid UTF-8 characters exists in the SQLite database.

## My opinion
Fuck it... idk.... Everything else technically disappears with the users, since tweets etc. tries to get them. e.g.
![image](https://user-images.githubusercontent.com/11597762/157478456-6a4e6d1b-f110-43ad-bf94-bffdaddeac4d.png)

**Comment by ValarMarkhulis** [comment]
Since the simulator is not logging in with the users, could we just delete the pw hashes and insert a valid hash string instead? The users are never gonna be used, because we don't know any of the original passwords anyway?

**Comment by Turboen** [comment]
> Since the simulator is not logging in with the users, could we just delete the pw hashes and insert a whatever valid pw hash as a byte[] instead? The users are never gonna be used, because we don't know any of the original passwords anyway?

Synes godt det kunne være en fin løsning.
Det eneste er dog, hvordan vi skal kunne argumentere for det i en "real life" situation, hvis han kunne finde på at spørge ind til det.

**Comment by ValarMarkhulis** [comment]
Haha, I would say it was a "routine security upgrade" , so the users need to update their password to a newer and more secure one.

**Comment by Thomsen96** [comment]
> > Since the simulator is not logging in with the users, could we just delete the pw hashes and insert a whatever valid pw hash as a byte[] instead? The users are never gonna be used, because we don't know any of the original passwords anyway?
> 
> Feel this could be a good enough solution for now. The only problem I see, is if we were to argue for doing it in a "real life" situation, if he might ask about at the exam.

I share the outlook of @Turboen. 

In this situation where the simulator is creating the information and never logging in again. Could we argue that a "real life" scenario is not realistic. 

In a real world scenario, could we just do the task over a longer period of time. Where when users are logging in, could we verify the password with the current method and then create a new column with an updated method of storing/data type that we could transfer. And when we have enough users that have been through the platform we could transition to the new database. And maybe keep the old one around for those that at this point hadn't used the platform in a while and then move the them when they do access the platform. 

But since we would never have the same user come back to the platfrom and login is would this never happen. and clean slate is more easy

**Comment by Zavir** [comment]
I think the argument proposed by @Thomsen96 is pretty good if we were to be questioned about it in an exam situation.


> Since the simulator is not logging in with the users, could we just delete the pw hashes and insert a whatever valid pw hash as a byte[] instead? The users are never going to be used, because we don't know any of the original passwords anyway?

I also agree with this solution.  "routine security upgrade" is also a nice explanation for the exam haha.
